  <launch>
    <arg name="cars_num" value="1" />
    <arg name="car_id" value="0"/>
    <arg name="car_width" value="0.5"/>
    <arg name="car_length" value="0.5"/>
    
    <arg name="init_x" value="3.05"/>
    <arg name="init_y" value="0.5"/>
    <arg name="init_z" value="2.8"/>  
    <arg name="init_yaw" value="0.0"/>

    <arg name="target_x" value="32.0"/>
    <arg name="target_y" value="0.0"/>
    <arg name="target_yaw" value="0.0"/>

    <arg name="odom_topic" value="/car_0/odometry"/>
    <arg name="lidar_topic" value="/car_0_laser_simulator/car_0_local_pointcloud"/>

    <arg name="cmd_topic" value="/car_0/cmd"/>
    <arg name="traj_topic" value="/car_0/minco_traj"/>

    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find plan_manage)/launch/multi_sim.rviz" required="true" />
    <!-- laser simulator -->
    <node pkg="laser_simulator" type="laser_sim_node" name="car_$(arg car_id)_laser_simulator" output="screen">
        <param name="car_id" value="$(arg car_id)" />
        <param name="map_topic" value="/global_cloud"/>
        <param name="odom_topic" value="$(arg odom_topic)"/>
        <param name="laser_pcd_topic" value="car_$(arg car_id)_local_pointcloud"/>
        <param name="laser_range_topic" value="car_$(arg car_id)_laser_range" />
        <param name="sensing_horizon" value="10.0"/>
        <param name="sensing_rate" value="10"/>
        <param name="pc_resolution" value="0.05"/>
        <param name="vtc_laser_line_num" value="32"/>
        <param name="vtc_laser_range_dgr" value="30.0"/>
        <param name="hrz_limited" value="false"/>
        <param name="hrz_laser_range_dgr" value="360.0"/>
        <param name="hrz_laser_line_num" value="360"/>
        <param name="use_resolution_filter" value="true"/>
    </node>

    <!-- kimatic simulator -->
    <node pkg="kimatic_simulator" name="car_$(arg car_id)_ugv_kimatic_simulator" type="kimatic_simulator_node" output="screen">
        <!-- <param name="map_name" value="RockyDesertMesh.pcd" /> -->
        <param name="map_name" value="processed_map_1.pcd" />
        <param name="max_height" value="0.05" />  <!--0.3-->
        <param name="init_x" value="$(arg init_x)"/>
        <param name="init_y" value="$(arg init_y)"/>
        <param name="init_z" value="$(arg init_z)"/>
        <param name="init_yaw" value="$(arg init_yaw)"/>
        <remap from="~command" to="$(arg cmd_topic)"/>
        <remap from="~odometry" to="$(arg odom_topic)"/>
        <remap from="~marker" to="$(arg odom_topic)_marker"/>
    </node>
    
    <!-- mpc controller -->
    <rosparam command="load" file="$(find mpc_controller)/params/controller.yaml" />
    <node pkg="mpc_controller" name="car_$(arg car_id)_mpc_controller_node" type="mpc_controller_node" output="screen" >
        <remap from="cmd" to="$(arg cmd_topic)"/>
        <remap from="odom" to="$(arg odom_topic)"/>
        <remap from="traj" to="$(arg traj_topic)"/>
        <remap from="traj" to="/planner/trajectory"/>
        <remap from="dptraj" to="/planner/dptrajectory"/>
    </node>  

    <!-- terrain analyzer -->
    <rosparam command="load" file="$(find terrain_analyzer)/params/elevation_map_gpu.yaml" />
    <node pkg="terrain_analyzer" name="car_$(arg car_id)_terrain_analyzer" type="terrain_analyzer" output="screen" >
        <remap from="~odom" to="$(arg odom_topic)"/>
        <remap from="~cloud" to="$(arg lidar_topic)/map"/>
    </node>

    <!-- main algorithm params -->
    <include file="$(find plan_manage)/launch/advanced_param.xml">

        <arg name="cars_num" value="$(arg cars_num)"/>
        <arg name="car_id" value="$(arg car_id)"/>
        <arg name="car_width" value="$(arg car_width)"/>
        <arg name="car_length" value="$(arg car_length)"/>

        <arg name="target_x" value="$(arg target_x)" />
        <arg name="target_y" value="$(arg target_y)" />
        <arg name="target_yaw" value="$(arg target_yaw)" />
        <arg name="in_swarm" value="true" />

        <arg name="odometry_topic" value="$(arg odom_topic)"/>
        <arg name="traj_topic" value="$(arg traj_topic)"/>

        <!-- maximum velocity and acceleration the car will reach -->
        <arg name="max_vel" value="2.0" />
        <arg name="max_acc" value="0.5" />
        <arg name="max_cur" value="0"/>
    </include>

  </launch>
